### ISSUE TITLE
Switch `EpochTime` to use `Int`

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/base#43]
Allowing it to talk about times before 1970.

**Comment by runarorama** [comment]
This needs to go with a change to a couple of builtins:

* `io.getFileTimestamp.impl`
* `io.systemTime.impl`

Thes currently return `Nat` but need to switch to return `Int`. The underlying Haskell functions return an `Int64` anyway so this should not be too hard.

**Comment by pchiusano** [comment]
`systemTimeMicroseconds` already returns `Int`. `systemTime` (which returns `Nat`) should just be deleted from base at this point... it's been in a couple releases even after `systemTimeMicroseconds` was introduced.

Unfortunate about the `getFileTimestamp` oversight, but yeah, that builtin should be changed (rather, we should introduce a new builtin that returns an `Int`, then delete the old).

**Comment by runarorama** [comment]
I've changed `EpochTime` to carry an `Int` instead of a `Nat`, but left the resolution the same (it's still 1-second resolution). We need to do a larger overhaul of dates, times, and clocks in general. But for now...

  The changes summarized below are available for you to review, using the following
  command:

    pull-request.load https://github.com/unisonweb/base:v3:.trunk https://github.com/runarorama/ubase2:.epochInt

  Updates:

     unique type io.EpochTime

     ↓
     unique type io.EpochTime
       +  unisoncomputing2022 : License
       +  runarorama          : Author

     io.EpochTime.EpochTime : Nat -> base.io.EpochTime
     ↓
     io.EpochTime.EpochTime : Int -> io.EpochTime

     io.getFileTimestamp : FilePath ->{IO, Exception} base.io.EpochTime
     ↓
     io.getFileTimestamp : FilePath ->{IO, Exception} io.EpochTime
     +  unisoncomputing2022 : License
     +  runarorama          : Author
     +  unisoncomputing2020 : License

     io.systemTime : '{IO, Exception} base.io.EpochTime
     ↓
     io.systemTime : '{IO} io.EpochTime
     +  unisoncomputing2022 : License
     +  runarorama          : Author
     +  unisoncomputing2020 : License

    There were 1 auto-propagated updates.

     patch patch (added 3 updates)

**Comment by atacratic** [comment]
Could we not go the whole way with this and switch to microseconds, as per the plan [here](https://github.com/unisonweb/unison/issues/2359#issuecomment-914366921)?  As it is, epochtime is kind of useless for many applications because of the low resolution.  For example for my datetime lib I would still need a translation to my own data structure, which is a papercut on every interaction with the library.  I think we could safely settle on microseconds without a full think-through of the date/time/clock space.

**Comment by runarorama** [comment]
I'm reluctant to switch the meaning of `EpochTime` as a patch, but I'm planning a better and upgraded clock API with nanosecond resolution:

```
unique type Clock = Monotonic | RealTime | CPUTime | ThreadTime

-- builtin
io.getTime : Clock -> {IO} Duration

io.getSystemTime : '{IO} Duration
io.getSystemTime _ = io.getTime RealTime

unique type Duration = { toSeconds : Int, nanoPortion : Nat }

Duration.toNanoseconds : Duration -> Int
Duration.toNanoseconds = cases
  Duration s n -> s * +1000000000 + n

Duration.toMicroseconds : Duration -> Int
Duration.toMicroseconds = cases
  Duration s n -> s * +1000000 + (n / 1000)

Duration.toMilliseconds : Duration -> Int
Duration.toMilliseconds = cases
  Duration s n -> s * +1000 + (n / 1000000)
```

`Duration` has as much resolution as you might ever want and also can represent times that are ~40x the Hubble time.

**Comment by atacratic** [comment]
Nice 👍

(throwaway remarks:  I wonder about `Duration` maybe carrying a resolution indication as a phantom type param.  Probably not.  Also the name 'duration' is maybe a bit awkward for negative values.  But TimeOffset or TimeIncrement or similar is ungainly.  I think I used 'interval' in my lib.)

**Comment by pchiusano** [comment]
@runarorama that API seems kinda complicated for a builtin, I'd see if you can provide the nano precision without having to introduce new types for it. 

I also find it weird to have potentially negative "durations". Duration would typically mean a distance between points in time, which is always positive, unless you're talking adding a direction (forwards or backwards). Which IMO makes it something other than a duration. Though maybe this is just a naming thing.

What about something like: `io.elapsedNanoseconds : Int ->{IO} Int` (where the input `Int` is `systemTimeMicroseconds`). So you can get nanosecond precision for computing deltas and you can build a library like what you propose.

```
-- probably fine for many purposes, for the next couple hundred years until int nanoseconds overflow
io.systemTimeNanoseconds : '{IO} Int
io.systemTimeNanoseconds _ = io.elapsedNanoseconds 0
```

Also, I am not a fan of introducing a random enum just so that we can overload the name `getTime`. It's slower since the implementation has to dispatch on that value, and it's just sort of a nonstandard calling convention. Just have `getTime.thread`, `getTime.wallClock` if you're going to go that route, and the enum thing if anyone wants that can be done with a user-level library rather than the builtin.

**Comment by runarorama** [comment]
What I've settled on is four builtins which expose Haskell's `System.Clock` API:

```
builtin type TimeSpec
builtin Clock.internal.getTime.v1 : Nat ->{IO} TimeSpec
builtin Clock.internal.sec.v1 : TimeSpec -> Int
builtin Clock.internal.nsec.v1 : TimeSpec -> Nat
```

`Duration` is maybe not a good name for a time offset. But the question "how long has it been since midnight UTC on Jan 1 1970?" is sometimes negative. We can bikeshed about the name. Haskell's System.Clock calls this `TimeSpec`. JDK 8 calls it `Duration` (and they can meaningfully be negative). I think it's valuable to have a single type that represents an amount of time elapsed, rather than having lots of `Int`, `Nat`, and `Float` meaning different things all over the place.

We can skip the enum since `getTime.v1` just takes a `Nat` (which is translated to an enum on the Haskell side).

**Comment by pchiusano** [comment]
Maybe `TimeOffset` is a good name for the concept of a `Duration` that can be negative?

I still dislike the single `getTime` with an enum for the reasons I gave earlier. Can you explain why you want to keep that? Just have separate builtins for fetching a `TimeOffset` with different meanings. Even as an implementation detail, there is no need to use a `Nat` as a pseudo-enum just because the Haskell API does this. IMO, System.Clock is a needlessly weird Haskell API that we don't need to replicate in Unison.

Furthermore, doing it with separate builtins means we don't need to add all the variants at once. I'm not even sure what all those modes mean and if they can all be easily supported by other runtimes, like the JIT compiled runtime we're working toward.

**Comment by runarorama** [comment]
Well, I don't want users to have to call `internal.getTime.v1`, for sure. They'll call one of e.g.:

```
Clock.now : '{IO, Exception} Instant
Clock.timeSinceEpoch : '{IO, Exception} Duration
Clock.monotonic : '{IO, Exception} Duration
Clock.processCPUTime : '{IO, Exception} Duration
Clock.processThreadTime : '{IO, Exception} Duration
```

Something like that. Should we rather make _those_ built-in?

**Comment by pchiusano** [comment]
> Something like that. Should we rather make those built-in?

Ya, that's what I was thinking.

**Comment by runarorama** [comment]
This is addressed in #94
